Medipen Reagent Examine Text Remaster 2: The Electric Boogaloo

567Turtle opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Adds a new, allergic reaction reducing, description to medipens, as well as a remastered petty_string for reagents which makes it much more modular.

## Why It's Good For The Game

Now, everybody hates when they take the allergy quirk and say, "Oh this pen is surely safe to use with my allergy, right?", then being on the verge of crit and popping their epi-pen, only to have a vicious allergic reaction do to the 3u of formaldehyde contained inside, who would've known that the epinephrine pen contained something that wasn't epinephrine? Shocking, I know. This update changes that, now before popping your medipen, the advanced scanners inside the pen (recycled pregnancy test parts) will scan the chemicals contained inside the pen and display them on the side of the pen. The hypospray chassis and combat medipens have not received the same remodel, due to it being far too expensive for the aforementioned models.
![image](https://user-images.githubusercontent.com/91508746/188498098-453fd174-dc3e-464c-939a-6a926413acd6.png)

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
qol: Changes how the medipens examine text works
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

567Turtle committed:
new medipen examine, petty_string remaster

Adds a new, allergic reaction reducing description to the medipen, as well as a remastered petty_string for reagents which makes it much more modular.

optimumtact commented:
I closed this already, shove off

567Turtle commented:
> I closed this already, shove off

But you closed it for such a shit reason, I put a lot of effort into this PR and to have you swoop in when it's finally all but perfect to close it because of a throwaway sentence I put in the description? What the fuck is your issue?

thgvr commented:
New player qol probably shouldn't be closed, there is precedent for showing reagents on chem applicators with patches already.

SuperNovaa41 commented:
> > I closed this already, shove off
> 
> But you closed it for such a shit reason, I put a lot of effort into this PR and to have you swoop in when it's finally all but perfect to close it because of a throwaway sentence I put in the description? What the fuck is your issue?

A maintainer can close PRs they don't like, not to mention headcoders especially

optimumtact commented:
> New player qol probably shouldn't be closed, there is precedent for showing reagents on chem applicators with patches already.

there's precedent for you shutting the fuck up too but you don't do that

Wallemations commented:
> > I closed this already, shove off
> 
> But you closed it for such a shit reason, I put a lot of effort into this PR and to have you swoop in when it's finally all but perfect to close it because of a throwaway sentence I put in the description? What the fuck is your issue?

While I agree he's a bit rude about it, the action bot clearly says that a maintainer can close your PR at will due to your low GBP score. There isn't really anything you can do about it from what I can tell, so I'd personally try to find a different thing to PR to get your GBP up, and maybe try again in the future once you have the GBP.

567Turtle commented:
> > > I closed this already, shove off
> > 
> > 
> > But you closed it for such a shit reason, I put a lot of effort into this PR and to have you swoop in when it's finally all but perfect to close it because of a throwaway sentence I put in the description? What the fuck is your issue?
> 
> While I agree he's a bit rude about it, the action bot clearly says that a maintainer can close your PR at will due to your low GBP score. There isn't really anything you can do about it from what I can tell, so I'd personally try to find a different thing to PR to get your GBP up, and maybe try again in the future once you have the GBP.

What do you think I was trying to do here? When I was just getting into dream maker I made a PR which was too big for my britches and I couldn't follow through with, my GBP got lowered by 4 putting me at -3. I took on this suggestion because I thought it'd be easy points, but then I started enjoying myself and had a lot of fun with it, spending 3 or 4 weeks fine tuning it and improving it to the point of panacea where it stands today, just having, and I put this politely as I can, the insufferable dick that oranges is roll in while everybody I've talked to, contributors, players, maintainers, other headcoders even say it's a good PR, he just rolls up and closes it because I made a throwaway comment about medical allergy being free points. He didn't close it because the code was bad, because the change was bad, because people didn't want it, he closed it because I made a small dig at something I was effecting with my PR.

Palpetinus commented:
Reward the chad Terry powerplayer for diving the code and skimming the wiki

Punish the virgin new player that doesn't know what's in pens yet

This is good and fine and we love it, I HATE new players and gameplay clarity!

optimumtact commented:
>everybody I've talked to, contributors, players, maintainers, other headcoders even say it's a good PR

why don't you get one of them to come and reopen the PR then huh? if they actually exist

Mothblocks commented on code/modules/reagents/reagent_containers/hypospray.dm:
```diff
@@ -138,10 +139,16 @@
 	. = ..()
-	if(reagents?.reagent_list.len)
-		. += span_notice("It is currently loaded.")
+	if (old_text)     //if old_text = true have the old examine text, if otherwise have the new examine text
```
```suggestion
	if (old_text)
```

Mothblocks commented on code/modules/reagents/chemistry/reagents.dm:
```diff
@@ -245,12 +245,37 @@ Primarily used in reagents/reaction_agents
+ * * capitalize_names will result in
+ * * Water, 5 | Silicon, 6 | Soup, 4 | Space lube, 8
+ *
+ * * * use (reagent_list, TRUE, ", ", TRUE, TRUE) for the formatting
```
Use named parameters here

Mothblocks commented on code/modules/reagents/chemistry/reagents.dm:
```diff
@@ -245,12 +245,37 @@ Primarily used in reagents/reaction_agents
+ * * * use (reagent_list, TRUE, ", ", TRUE, TRUE) for the formatting
+ * * * Water, Silicon, Soup, and Space Lube
+ */
+/proc/pretty_string_from_reagent_list(list/reagent_list, names_only, join_text = " | ", final_and, capitalize_names)
```
You can also skip the named parameters part entirely if `names_only`, `final_and`, and `capitalize_names` was just all one bitfield

Mothblocks commented on code/modules/reagents/chemistry/reagents.dm:
```diff
@@ -245,12 +245,37 @@ Primarily used in reagents/reaction_agents
+ */
+/proc/pretty_string_from_reagent_list(list/reagent_list, names_only, join_text = " | ", final_and, capitalize_names)
 	//Convert reagent list to a printable string for logging etc
 	var/list/rs = list()
```
Rename to `reagent_strings` while you're here

Mothblocks replied:
This hasn't been fixed

567Turtle replied:
Coulda swore I fixed all that, I'll do that now

Mothblocks commented on code/modules/reagents/chemistry/reagents.dm:
```diff
@@ -245,12 +245,37 @@ Primarily used in reagents/reaction_agents
 	var/list/rs = list()
+	var/reagents_left = reagent_list.len
+	var/intial_list_length = reagents_left
 	for (var/datum/reagent/R in reagent_list)
```
```suggestion
	for (var/datum/reagent/reagent as anything in reagent_list)
```

Then change other code

Mothblocks commented on code/modules/reagents/reagent_containers/hypospray.dm:
```diff
@@ -116,6 +116,7 @@
 	list_reagents = list(/datum/reagent/medicine/epinephrine = 10, /datum/reagent/toxin/formaldehyde = 3, /datum/reagent/medicine/coagulant = 2)
 	custom_price = PAYCHECK_CREW
 	custom_premium_price = PAYCHECK_COMMAND
+	var/old_text
```
Also this name is bad. It should more clearly articulate what it does. "Old" text isn't going to mean anything in a year or so when nobody remembers what it is.

Mothblocks commented on code/modules/reagents/reagent_containers/hypospray.dm:
```diff
@@ -138,10 +139,16 @@
 	else
-		. += span_notice("It is spent.")
+		if(length(reagents?.reagent_list))
+			. += span_notice("There\'s a small LCD screen on the side of the medipen which reads, 'WARNING: This medipen contains [pretty_string_from_reagent_list(reagents.reagent_list, TRUE, ", ", TRUE, TRUE)]. Do not use if allergic to any listed chemicals.' in small text.") //cheapskates at nanotrasen couldnt ball out for LED
```
Instead of using the exact reagents list, this should instead be generated in `Initialize` and represent what the default state of the reagents are. You can still have the "It is spent" text, but I don't think it should be dynamic, especially for something like CMO hypospray.

567Turtle replied:
Hypospray and combat pens don't get the new text

Mothblocks replied:
I think they ought to since it's nice to be able to tell people hyposprays have omnizine and such through the same system. I just don't think it should auto update.

567Turtle replied:
I think they shouldnt because as stated in the past iteration of this PR, "reduces espionage opportunity", you cant put custom reagents in a medipen, you can put custom reagents in a hypospray.

Mothblocks replied:
But that's why I made the suggestion:

> Instead of using the exact reagents list, this should instead be generated in `Initialize` and represent what the default state of the reagents are. You can still have the "It is spent" text, but I don't think it should be dynamic, especially for something like CMO hypospray.

567Turtle replied:
Hypospray bottles are marked and labelled already, there wouldnt be any purpose in having the initialize chemicals displayed, as they would just say "Airuri" or whatever the default chemical in the hypospray is, when it could easily be deducted that Airuri is inside of the hypospray, just by ejecting the cartridge.

SexmasterWasTaken commented:
> > everybody I've talked to, contributors, players, maintainers, other headcoders even say it's a good PR
> 
> why don't you get one of them to come and reopen the PR then huh? if they actually exist

lol

567Turtle committed:
mothblocks suggestions

I thought it was really funny when oranges tried to call my bluff that I had talked to people about my code change being good and then was immediately disproven, then he just went silent

567Turtle committed:
probably fixes the things

spells out the RS in reagents.dm where I was digging around and reverts an obligatory change I made mistakingly

567Turtle commented:
wait did a fuck up let me fix that

567Turtle committed:
fixed thing

forgot to change other rs', fixed it now

optimumtact commented:
IM SHRINKING, AHHHHHHHHHHH IM MELTING

567Turtle commented:
> IM SHRINKING, AHHHHHHHHHHH IM MELTING

what did you mean by this

Mothblocks commented:
Being in discord shrinks your power level

LemonInTheDark commented:
impeccable timing my friend

Mothblocks commented:
I was checking notifications while waiting for a 20-30 minute compile time at lunch

567Turtle commented:
I dunno what you people are talking about but.... yea
